lib/rust/prelude/src/data/vec_indexed_by.rs

```diff
@@ -0,0 +1,415 @@
+            deserialize = "Vec<T, A>: crate::serde_reexports::Deserialize<'de>"
+        ))
+    )]
+    vec: Vec<T, A>,
+    key: PhantomData<I>,
+}
+
+impl<T, I> VecIndexedBy<T, I> {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Vec::with_capacity(capacity).into()
+    }
+}
+
+impl<T, I, A> VecIndexedBy<T, I, A>
+where A: Allocator
+{
+    /// Length of the vector.
+    pub fn len(&self) -> usize {
+        self.vec.len()
+    }
+
+    /// Check if the vector is empty.
+    pub fn is_empty(&self) -> bool {
+        self.vec.is_empty()
+    }
+
+    /// Returns the number of elements the vector can hold without reallocating.
+    pub fn capacity(&self) -> usize {
+        self.vec.capacity()
+    }
+
+    /// Push a new element to the vector.
+    pub fn push(&mut self, value: T) {
+        self.vec.push(value)
+    }
+
+    /// Extend the vector with new elements.
+    pub fn extend<Iter: IntoIterator<Item = T>>(&mut self, iter: Iter) {
+        self.vec.extend(iter)
+    }
+
+    /// Reserves capacity for at least additional more elements to be inserted in the given vector.
+    pub fn reserve(&mut self, additional: usize) {
+        self.vec.reserve(additional)
+    }
+
+    /// Shrinks the capacity of the vector as much as possible.
+    pub fn shrink_to_fit(&mut self) {
+        self.vec.shrink_to_fit()
+    }
+
+    /// Removes the last element from a vector and returns it, or [`None`] if it is empty.
+    pub fn pop(&mut self) -> Option<T> {
+        self.vec.pop()
+    }
+
+    /// Returns the first element of the slice, or [`None`] if it is empty.
+    pub fn first(&self) -> Option<&T> {
+        self.vec.first()
+    }
+
+    /// Returns the last element of the slice, or [`None`] if it is empty.
+    pub fn last(&self) -> Option<&T> {
+        self.vec.last()
+    }
+
+    /// Returns a mutable pointer to the first element of the slice, or [`None`] if it is empty.
+    pub fn first_mut(&mut self) -> Option<&mut T> {
+        self.vec.first_mut()
+    }
+
+    /// Returns a mutable pointer to the last element of the slice, or [`None`] if it is empty.
+    pub fn last_mut(&mut self) -> Option<&mut T> {
+        self.vec.last_mut()
+    }
+
+    /// Returns an iterator over the slice.
+    pub fn iter(&self) -> slice::Iter<'_, T> {
+        self.vec.iter()
+    }
+
+    /// Returns a mutable iterator over the slice.
+    pub fn iter_mut(&mut self) -> slice::IterMut<'_, T> {
+        self.vec.iter_mut()
+    }
+
+    /// Resizes the Vec in-place so that len is equal to new_len.
+    pub fn resize_with(&mut self, new_len: usize, f: impl FnMut() -> T) {
+        self.vec.resize_with(new_len, f)
+    }
+
+    /// Shortens the vector, keeping the first len elements and dropping the rest.
+    pub fn truncate(&mut self, len: usize) {
+        self.vec.truncate(len)
+    }
+
+    /// Extracts a slice containing the entire vector.
+    pub fn as_slice(&self) -> &[T] {
+        self.vec.as_slice()
+    }
+}
```

Comment:
Because I don't want to inherit by accident methods that would allow use number-based indexes instead of using custom index type. But maybe we can just do it and Fix it when we will find such. method. Yeah, lets do it.